File: fsharp/triangle/Triangle.fs
module Triangle

let private eval (lst: float list, i: int) =
    let processLst (l: float list) =
        let equality (l: float list): bool = l.Item 0 + l.Item 1 >= l.Item 2 && l.Item 1 + l.Item 2 >= l.Item 0 && l.Item 2 + l.Item 0 >= l.Item 1

        if not (equality lst) then None
        else
            match List.distinct l with 
            | x when (List.exists ((=) 0.0) x) -> None
            | x -> Some x.Length

    match processLst lst with
    | Some x -> x = i
    | None -> false

let equilateral triangle = eval(triangle, 1)

let isosceles triangle = eval(triangle, 1) || eval(triangle, 2)

let scalene triangle = eval(triangle, 3)

File: csharp/triangle/Triangle.cs
using System;
using System.Linq;

public static class Triangle
{
    public enum Type : int
    {
        Equilateral = 0,
        Isosceles,
        Scalene
    }

    public static bool IsScalene(double side1, double side2, double side3)
        => ProcessTriangle(Type.Scalene, side1, side2, side3);

    public static bool IsIsosceles(double side1, double side2, double side3) 
        => ProcessTriangle(Type.Isosceles, side1, side2, side3);

    public static bool IsEquilateral(double side1, double side2, double side3)
        => ProcessTriangle(Type.Equilateral, side1, side2, side3);

    private static bool ProcessTriangle(Type expectation, params double[] sides)
    {
        if (!ValidateInput(sides)) return false;

        Type distance = (Type)HammingDistance(sides);

        if (expectation == Type.Isosceles)
            return distance == Type.Equilateral || distance == Type.Isosceles;

        return distance == expectation;
    }

    private static bool ValidateInput(params double[] sides)
    {
        if (sides.Length < 3) return false;

        if (!sides.All(s => s > 0)) return false;

        var windows = new[] {
            (sides[0], sides[1], sides[2]),
            (sides[1], sides[2], sides[0]),
            (sides[2], sides[0], sides[1])
        };
        
        return windows.All(t => t.Item1 + t.Item2 >= t.Item3);
    }

    public static int HammingDistance(params double[] sides)
    {
        var diffs = sides.Distinct();

        var x = diffs.Count() - 1;
        return x;
    }
    public static bool IsDegenerate(params double[] sides)
        => sides[0] + sides[1] == sides[2];
}